### DISCUSSION TITLE
How to handle postgres interval_style iso_8601 with sqlalchemy and psycopg

### DISCUSSION THREAD

**Discussion by Kaelten** [sqlalchemy/sqlalchemy#8274] [Usage Questions]
For a variety of reasons I've needed to set the default interval style on my database to `iso_8601`.  Unfortunately, psycopg has decided not to support this when querying intervals from the database.  As a result I get this error if I try:

```python
Traceback (most recent call last):
  File "/opt/app/.venv/lib/python3.10/site-packages/sqlalchemy/engine/cursor.py", line 977, in fetchall
     rows = dbapi_cursor.fetchall()
     psycopg2.NotSupportedError: iso_8601 intervalstyle currently not supported
```

I've tried multiple things to work around it and they all partially work.  First I created an engine connect hook:

```python
@event.listens_for(engine, 'connect')
def connection_variables(dbapi_connection: Any, connection_record: Any) -> None:  # noqa: ANN401
    # track our pid to ensure forked processes do not share connections
    connection_record.info['pid'] = os.getpid()

    # appears that psycopg doesn't have full implementation for iso_8601 styles, let's work with
    #   the default
    with dbapi_connection.cursor() as cursor:
        cursor.execute("SET intervalstyle = 'postgres';")
```

originally this was just the `cursor.execute` part which worked in most situations but I started having issues where it'd happen periodically in my workers.  I expanded this into using the connection record tracking to ensure we weren't sharing connections across forked processes by paring the above with this handler:

```python
@event.listens_for(engine, 'checkout')
def on_connection_checkout(
    dbapi_connection: Any,  # noqa: ANN401
    connection_record: Any,  # noqa: ANN401
    connection_proxy: Any,  # noqa: ANN401
) -> None:
    # invalidate the connection if it originated in another process. based on example here:
    #   https://docs.sqlalchemy.org/en/14/core/pooling.html#using-connection-pools-with-multiprocessing-or-os-fork
    if connection_record.info['pid'] != os.getpid():
        connection_record.dbapi_connection = connection_proxy.dbapi_connection = None

        message = f'connection belongs to another process ({connection_record.info["pid"]})'
        logging.warning(message)
        raise DisconnectionError(message)
```

However, this still didn't solve the problem reliabily.  As such I expanded the checkout event to also include this:

```python
    with dbapi_connection.cursor() as cursor:
        try:
            cursor.execute("SELECT interval '2 hours'")
        except NotSupportedError as e:
            connection_record.dbapi_connection = connection_proxy.dbapi_connection = None

            message = 'connection appears to be using iso_8601 interval style'
            logging.warning(message)
            raise DisconnectionError(message) from e
```

Despite this, I end up getting the original exception.  I've been unable to determine my exact reproduction case, but I believe it's likely related to connections in kubernetes being partially reestablished and sqlalchemy not knowing that it's a new connection.  

When this starts happening to the worker, it seems one or more of the connections will get stuck in this state and the above checkout code doesn't let me gracefully recover.  Ultimately, that's what I'm looking for is a way to reliably detect this situation, force sqlalchemy to reestablish the connection and execute the `SET intervalstyle` again. 

Any thoughts on this situation and how I could handle it better?

**Comment by zzzeek** [comment]
hey -

I'm not really following the second part where you're dealing with pid travelling and all that.   the connection's state can be set up front using the `@event.listens_for(engine, 'connect')` event as you have, and that would be for the life of the connection on that session.     there's other places to set it.  if you are using pgbouncer that would make these things more tricky but i dont know what kubernetes would be doing that would be affecting this unless it is also implementing some kind of postgresql connection scoping.

**Reply by Kaelten** [reply]
I'm not certain what's going on either honestly.  

Engine config looks like this:

```python
engine = create_engine(
    get_settings().database_dsn,
    future=True,
    pool_size=20,
    pool_pre_ping=True,
    pool_use_lifo=True,
    echo=get_settings().debug_mode,
)
```

And I'm only getting the issue on some connections inside of my staging and production environments.  In those environments sqlalchemy is connecting to the database cluster through the kubernetes in-cluster networking to a cluster of servers powered by https://github.com/zalando/postgres-operator through a service reference.  

I'm not aware of pg pool our bouncer being involved, but some of this is still a bit of black box.  That said, I don't expect you to help me debug my environment, but I'm pretty surprised that the checkout hook doesn't heal this gracefully.  Best I can tell the connection must pass that check and then somehow the connection looses that state by the time the session is actually used.

**Reply by zzzeek** [reply]
so possibilities include:

1. something else is using create_engine elsewhere  (can be application level)
2. something is calling a command like [DISCARD](https://www.postgresql.org/docs/current/sql-discard.html) or other PostgreSQL session-resetting command  (can be application level, or if some kind of cross-session PG connection scoping is occurring as occurs within pgbouncer)

I would add more debugging and logging.  Add assertions that check the desired "interval" setting on the connection for every checkout using the "checkout" event (of course you could just set the interval setting on every checkout also but this seems like it's a workaround).

**Reply by Kaelten** [reply]
@zzzeek The codebase is entirely first party, at least anything touching sqlalchemy is, and there's nothing that would call create_engine anywhere else. 

One thing that is confusing to me:

```python

@event.listens_for(engine, 'checkout')
def on_connection_checkout(dbapi_connection: Any, connection_record: Any, connection_proxy: Any) -> None:
    with dbapi_connection.cursor() as cursor:
        try:
            cursor.execute("SELECT interval '2 hours'")
        except NotSupportedError as e:
            connection_record.dbapi_connection = connection_proxy.dbapi_connection = None

            message = 'connection appears to be using iso_8601 interval style'
            logging.warning(message)
            raise DisconnectionError(message) from e
```

I would expect that code to force a new connection to be created.  In practice, it doesn't seem to be.  Would that point to a specific direction?

**Reply by zzzeek** [reply]
it should do something like that, I know that pattern is documented but it's pretty old at this point, and I think overall it's much better to use approach 2 at https://docs.sqlalchemy.org/en/14/core/pooling.html#using-connection-pools-with-multiprocessing-or-os-fork , call engine.dispose() up front.     For the code you have above, I would mostly expect if it's failing, it's that the NotSupportedError you see there is either not being thrown (the cursor.execute() is in fact succeeding, since you are doing the SET elsewhere) or not being caught (the origin of NotSupportedError class is from one of two places, psycopg2 directly or SQLAlchemy, and those are different).

**Reply by Kaelten** [reply]
@zzzeek I've found out how to adopt the second method listed there.  So moved that over.

I've also updated the checkout event to be capture both the sqlalchemy and psycopg2 versions of the exception.  

I've tried to confirm that the error would trigger as well.  When I'm in an interactive shell I initialized a session object and ran the following:

```python

session.execute("SET intervalstyle = 'iso_8601';")
session.execute("SELECT interval '2 hours'").all()
```

doing this successfully triggers the exception:

```
sqlalchemy.exc.NotSupportedError: (psycopg2.NotSupportedError) iso_8601 intervalstyle currently not supported
```

Does this validate that the exception should trigger when I call `cursor.execute`?

**Reply by zzzeek** [reply]
yes, I wasn't questioning that the bad interval setting would trigger the exception.    that code is actually managing to not test any of the things I was asking :)

* the code you have is using cursor.execute().  That will throw a `psycopg2.NotSupportedError`, not a  `sqlalchemy.exc.NotSupportedError`.  This is in contrast to using `session.execute()` which will throw `sqlalchemy.exc.NotSupportedError`.  Your previous example doesn't indicate which NotSupportedError class is being caught, as written it would need to be the psycopg2 version.
* your previous example code is only calling "SELECT interval '2 hours'".  So it's not clear that this is actually failing at this point, and if the reason for the failures you are seeing is happening somewhere else (such as, some other part of the application has its own create_engine() somewhere without this event hook)

**Reply by Kaelten** [reply]
I updated the hook to capture both the SQLAlchemy version and the psycopg2 versions.  Updated code is below.  Unfortunately, I still have the issue occuring.

```python

@event.listens_for(engine, 'checkout')
def on_connection_checkout(
    dbapi_connection: Any,  # noqa: ANN401
    connection_record: Any,  # noqa: ANN401
    connection_proxy: Any,  # noqa: ANN401
) -> None:
    with dbapi_connection.cursor() as cursor:
        try:
            cursor.execute("SELECT interval '2 hours'")
        except (SQLAlchemyNotSupportedError, EngineNotSupportedError) as e:
            connection_record.dbapi_connection = connection_proxy.dbapi_connection = None

            message = 'connection appears to be using iso_8601 interval style'
            logging.warning(message)
            raise DisconnectionError(message) from e
```

There's no other engine declarations in the code base and I have no libraries that would generate one either.

**Reply by Kaelten** [reply]
For what it's worth, I do think there's something else weird going on in the whole setup, but it's not making a lot of senes what since the code side appears like it should be working fine.

I've thought about putting some middleware in my workers that would invalidate the connection on this exception but I'm not actually sure how I'd code that.

**Reply by zzzeek** [reply]
do you see the warning message you added being emitted?

**Reply by Kaelten** [reply]
Nope it's not being written out.  Not sure how it'd make sense that it'd work in the checkout hook but fail in the code a few milliseconds later. :/

**Reply by Kaelten** [reply]
I think I figured it out!

After adding a bunch of logging around I don't think the `cursor.execute()` call was reliably triggering the exception.  I added a `cursor.fetchone()` call after and now it seems to be reliabily detecting and fixing the issue! 

I'm going to test more but it looks promising!

**Reply by zzzeek** [reply]
OK.  it would be nice to know why your original SET statement is not sufficient.

**Reply by Kaelten** [reply]
> OK. it would be nice to know why your original SET statement is not sufficient.

I agree.  My best guess has to do with how kubernetes networking works, but I'm not 100% certain.

I say that because this whole thing is similar to situations I've seen with network proxies where only half of the connection is reestablished for whatever reason.  There's many software proxies involved in routing between pods, but I wouldn't expect the proxy-to-postgres part of the connection to reset transparently since there's nothing like bgbouncer involved.  

My best supporting evidence of this being a possible cause was when the postgres cluster would change leaders.  When that happens the kubernetes service resource would repoint to a different pod and I'd see a spike of these errors.  I don't consider this to be a smoking gun though, since the errors could occur anytime of the day, not just during maintenance windows.

The other theory that makes sense to me is somehow a connection reset command is firing, but if that was the case I'd expect more uniformity on how it shows up.  As it stands I only see this issue on a fraction of requests/jobs, about 10% under load.  I also saw that once a given thread started having the issue it'd get stuck that way for a while.  

It's a bit of a mystery :/
